### PR TITLE
fix dotnet#43523 HttpSys Content-Length and Transfer-Encoding conflicts

### LINKED ISSUE #43523: Content-Length and Transfer-Encoding conflicts
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

https://datatracker.ietf.org/doc/html/rfc7230#section-3.3.2
>    A sender MUST NOT send a Content-Length header field in any message
   that contains a Transfer-Encoding header field.

https://datatracker.ietf.org/doc/html/rfc7230#section-3.3.3
> If a message is received with both a Transfer-Encoding and a
       Content-Length header field, the Transfer-Encoding overrides the
       Content-Length.  Such a message might indicate an attempt to
       perform request smuggling ([Section 9.5](https://datatracker.ietf.org/doc/html/rfc7230#section-9.5)) or response splitting
       ([Section 9.4](https://datatracker.ietf.org/doc/html/rfc7230#section-9.4)) and ought to be handled as an error.  A sender MUST
       remove the received Content-Length field prior to forwarding such
       a message downstream.

Kestrel follows this today by honoring the `Transfer-Encoding: chunked` and ignoring the Content-Length header. However, the Content-Length header is still present on the request and apps may still be consuming it.
https://github.com/dotnet/aspnetcore/blob/94cc679bce4967568fd458433ecda3b3fb0578b6/src/Servers/Kestrel/Core/src/Internal/Http/Http1MessageBody.cs#L146-L177

E.g. YARP has to check for this condition too to avoid a smuggling attempt when proxying.
https://github.com/microsoft/reverse-proxy/blob/06f28a36140543512a3fc3e1e9423094c43296ec/src/ReverseProxy/Forwarder/HttpTransformer.cs#L80-L92

App developers are not usually aware of spec nuances like this and write logic that makes decisions based on the Content-Length header when present. This can lead to application level functional and security issues like over/under buffering.

### Expected Behavior

Proposal: Kestrel should implement the following spec guideline "A sender MUST remove the received Content-Length field prior to forwarding such a message downstream.", treating the application as the downstream entity. If there is a Transfer-Encoding header then kestrel should remove any Content-Length header to avoid app level confusion.

Since Kestrel avoids modifying the request in most cases, giving the app as much of the original data as possible, we want to preserve the Content-Length header value by moving it to `x-Content-Length`. This should avoid people consuming the unvalidated value by accident, but it's still available for compat reasons.

YARP could remove its mitigation after this change.

### Steps To Reproduce

```
POST / HTTP/1.1
Content-Length: 500000
Transfer-Encoding: chunked

5
Hello
0
```

### Exceptions (if any)

_No response_

### .NET Version

_No response_

### Anything else?

Recommended for .NET 8.

### PR DESCRIPTION
**Pull request by FatTigerWang**
# fix #43523 HttpSys Content-Length and Transfer-Encoding conflicts

- [x] You've read the [Contributor Guide](https://github.com/dotnet/aspnetcore/blob/main/CONTRIBUTING.md) and [Code of Conduct](https://github.com/dotnet/aspnetcore/blob/main/CODE-OF-CONDUCT.md).
- [x] You've included unit or integration tests for your change, where applicable.
- [x] You've included inline docs for your change, where applicable.
- [x] There's an open issue for the PR that you are making. If you'd like to propose a new feature or change, please open an issue to discuss the change or find an existing issue.

Summary of the changes (Less than 80 chars)

## Description

Fixes #43523 HttpSys Content-Length and Transfer-Encoding conflicts

### CODE REVIEW

**Comment by Tratcher** [comment]
This will need some unit tests.

Also, this part of the original issue is missing:
> This should probably be triggered more eagerly from the constructor so apps don't read the CL header directly before it's dropped.
https://github.com/dotnet/aspnetcore/blob/main/src/Servers/HttpSys/src/RequestProcessing/Request.cs#L36

**Comment by FatTigerWang** [comment]
> This should probably be triggered more eagerly from the constructor so apps don't read the CL header directly before it's dropped.
https://github.com/dotnet/aspnetcore/blob/main/src/Servers/HttpSys/src/RequestProcessing/Request.cs#L36

Can you explain it specifically? Should I check the CL header in the [constructor](https://github.com/dotnet/aspnetcore/blob/main/src/Servers/HttpSys/src/RequestProcessing/Request.cs#L97) and remove it instead of removing it on the ContentLegnth's get accessor?

**Comment by Tratcher** [comment]
The code you changed isn't run until someone accesses the Request.Body, right? You can test that in the debugger by setting a break point. That would allow an application to read the invalid Content-Length header text from the collection before your code runs to modify it. So we need to find a way to make that modification code run earlier, like calling it from the Request constructor.

**Comment by FatTigerWang** [comment]
I will add unit tests later.

**Comment by FatTigerWang** [comment]
I found this test to be invalid:
https://github.com/dotnet/aspnetcore/blob/main/src/Servers/HttpSys/test/FunctionalTests/RequestHeaderTests.cs#L58
It will pass the test even if the assert fails because the exception is caught and there is no assert response.

**Comment by FatTigerWang** [comment]
I want to add a unit test here like this [method](https://github.com/dotnet/aspnetcore/blob/main/src/Servers/IIS/IIS/test/Common.FunctionalTests/RequestResponseTests.cs#L333), but it always gets skipped.

File: src/Servers/HttpSys/test/FunctionalTests/RequestHeaderTests.cs [hunk 1]
```diff
@@ -114,4 +189,35 @@ private async Task SendRequestAsync(string address, string customHeader, string[
...
+
+        byte[] request = Encoding.ASCII.GetBytes(builder.ToString());
+
+        using (Socket socket = new Socket(SocketType.Stream, ProtocolType.Tcp))
+        {
+            socket.Connect(uri.Host, uri.Port);
+            socket.Send(request);
+            byte[] response = new byte[1024 * 5];
+            await Task.Run(() => socket.Receive(response));
+            return Encoding.ASCII.GetString(response);
```
Comment (Tratcher) [inline src/Servers/HttpSys/test/FunctionalTests/RequestHeaderTests.cs:220]: Only the filled part of the buffer should be decoded, Receive returns how much was read.
```suggestion
            var read = await Task.Run(() => socket.Receive(response));
            return Encoding.ASCII.GetString(response, 0, read);
```

**Comment by Tratcher** [comment]
> I want to add a unit test here like this [method](https://github.com/dotnet/aspnetcore/blob/main/src/Servers/IIS/IIS/test/Common.FunctionalTests/RequestResponseTests.cs#L333), but it always gets skipped.

Skipped? Note those tests are set up to run in several different projects and to skip if some of the dependencies aren't met. E.g. one project runs it for IIS, but only if installed and you're running as admin. The IIS Express test project should always run these on Windows.

**Comment by FatTigerWang** [comment]
The unit tests were skipped because I didn't install [ASP.NET Hosting bundle](https://docs.microsoft.com/en-us/aspnet/core/host-and-deploy/iis/hosting-bundle), and now the unit tests are working fine after the installation is complete.

**Review by Tratcher — APPROVED** [review]

**Comment by Tratcher** [comment]
Thanks

**Pull request merged**